src/middleware/ip-restriction/index.ts

```diff
@@ -0,0 +1,147 @@
+/**
+ * Middleware for restrict IP Address
+ * @module
+ */
+
+import type { Context, MiddlewareHandler } from '../..'
+import type { AddressType, GetConnInfo } from '../../helper/conninfo'
+import { createMiddleware } from '../../helper/factory'
+import { HTTPException } from '../../http-exception'
+import {
+  convertIPv4ToBinary,
+  convertIPv6ToBinary,
+  distinctRemoteAddr,
+  expandIPv6,
+} from '../../utils/ipaddr'
+
+/**
+ * Function to get IP Address
+ */
+type GetIPAddr = GetConnInfo | ((c: Context) => string)
+
+/**
+ * ### IPv4 and IPv6
+ * - `*` match all
+ *
+ * ### IPv4
+ * - `192.168.2.0` static
+ * - `192.168.2.0/24` CIDR Notation
+ *
+ * ### IPv6
+ * - `::1` static
+ * - `::1/10` CIDR Notation
+ */
+export type IPRestrictRule = string | ((addr: { addr: string; type: AddressType }) => boolean)
+
+const IS_CIDR_NOTATION_REGEX = /\/[0-9]{0,3}$/
+export const isMatchForRule = (
+  remote: {
+    addr: string
+    type: AddressType
+  },
+  rule: IPRestrictRule
+): boolean => {
+  if (rule === '*') {
+    // Match all
+    return true
+  }
+  if (typeof rule === 'function') {
+    return rule(remote)
+  }
+  if (IS_CIDR_NOTATION_REGEX.test(rule) && (remote.type === 'IPv4' || remote.type === 'IPv6')) {
+    const isIPv4 = remote.type === 'IPv4'
+
+    const splitedRule = rule.split('/')
+    // CIDR
+    const baseRuleAddr = splitedRule[0]
+    if (distinctRemoteAddr(baseRuleAddr) !== remote.type) {
+      return false
+    }
+    const prefix = parseInt(splitedRule[1])
+
+    const addrToBinary = isIPv4 ? convertIPv4ToBinary : convertIPv6ToBinary
+
+    const baseRuleMask = addrToBinary(baseRuleAddr)
+    const remoteMask = addrToBinary(remote.addr)
+    const mask = ((1n << BigInt(prefix)) - 1n) << BigInt((isIPv4 ? 32 : 128) - prefix)
+
+    return (remoteMask & mask) === (baseRuleMask & mask)
+  }
+
+  const ruleAddrConnType = distinctRemoteAddr(rule)
+  if (ruleAddrConnType === 'IPv4' || ruleAddrConnType === 'IPv6') {
+    // Static
+    if (ruleAddrConnType !== remote.type) {
+      return false
+    }
+    if (remote.type === 'IPv6') {
+      return expandIPv6(remote.addr) === expandIPv6(rule)
+    }
+    return rule === remote.addr // IPv4
+  }
+  throw new TypeError('Rule is unknown')
+}
+
+/**
+ * Rules for IP Limit Middleware
+ */
+export interface IPRestrictRules {
+  denyList?: IPRestrictRule[]
+  allowList?: IPRestrictRule[]
+}
+
+/**
+ * IP Limit Middleware
+ *
+ * @param getIP function to get IP Address
+ */
+export const ipRestriction = (
+  getIP: GetIPAddr,
+  { denyList = [], allowList = [] }: IPRestrictRules,
+  onError?: (remote: { addr: string; type: AddressType }) => Response | Promise<Response>
+): MiddlewareHandler => {
+  const denyLength = denyList.length
+  const allowLength = allowList.length
+
+  const blockError = (): HTTPException =>
+    new HTTPException(403, {
+      res: new Response('Unauthorized', {
+        status: 403,
+      }),
+    })
+
+  return createMiddleware(async (c, next) => {
```

Comment:
It is sometimes helpful to be able to display the middleware name when debugging, so we use named functions for functions included in the hono repository.

```suggestion
  return async function ipRestriction(c, next) {
```

Reply:
I agreed! I changed.

Reply:
👍